lib/error_report_notify.c

```diff
@@ -0,0 +1,225 @@
+	 */
+	if (atomic_cmpxchg_acquire(&current_pid, -1, get_encoded_pid()) != -1)
+		return;
+	current_report_pos = 0;
+}
+
+/*
+ * Trace hook for the error_report_end event. If an event from the mismatching
+ * error_report_start is received, it is ignored. Otherwise, null-terminate the
+ * buffer, increase the report count (effectively releasing the report to
+ * last_report_show() and schedule a notification about a new report.
+ */
+static void probe_report_end(void *ignore, const char *tool_name,
+			     unsigned long id)
+{
+	pid_t pid = atomic_read(&current_pid);
+	int idx;
+
+	if (pid != get_encoded_pid())
+		return;
+
+	idx = (atomic_read(&num_reports) + 1) % 2;
+	if (current_report_pos == BUF_SIZE)
+		report_buffer[idx][current_report_pos - 1] = 0;
+	else
+		report_buffer[idx][current_report_pos] = 0;
+
+	/* Pairs with acquire in last_report_show(). */
+	atomic_inc_return_release(&num_reports);
+	schedule_delayed_work(&reporting_done, 0);
+	/*
+	 * Release the writer lock. Pairs with the acquire in
+	 * probe_report_start().
+	 */
+	atomic_set_release(&current_pid, -1);
+}
+
+/*
+ * Trace hook for the console event. If a line comes from a task/CPU that did
+ * not send the error_report_start event, that line is ignored. Otherwise, it
+ * is stored in the report_buffer[(num_reports + 1) % 2].
+ */
+static void probe_console(void *ignore, const char *buf, size_t len)
+{
+	int pid = atomic_read(&current_pid);
+	size_t to_copy;
+	int idx;
+
+	if (pid != get_encoded_pid() || truncated)
+		return;
+
+	idx = (atomic_read(&num_reports) + 1) % 2;
+	to_copy = min(len, BUF_SIZE - current_report_pos);
+	memcpy(report_buffer[idx] + current_report_pos, buf, to_copy);
+	current_report_pos += to_copy;
+	if (len > to_copy) {
+		truncated = true;
+		memcpy(report_buffer[idx] + current_report_pos - sizeof(TRUNC_MSG),
+		       TRUNC_MSG, sizeof(TRUNC_MSG));
+	}
+}
+
+static void register_tracepoints(void)
+{
+	register_trace_console(probe_console, NULL);
+	register_trace_error_report_start(probe_report_start, NULL);
+	register_trace_error_report_end(probe_report_end, NULL);
+}
+
+/*
+ * read() handler for /sys/kernel/error_report/last_report.
+ * Because the number of reports can change under our feet, check it again
+ * after copying the report, and retry if the numbers mismatch.
+ * */
+static ssize_t last_report_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	/* Pairs with release in probe_report_end(). */
+	int index = atomic_read_acquire(&num_reports), old_index;
+	ssize_t ret;
+
+	do {
+		old_index = index;
+		/*
+		 * If index and old_index mismatch, we might be accessing
+		 * report_buffer concurrently with a writer thread. In that
+		 * case the read data will be discarded.
+		 */
+		ret = data_race(strscpy(buf, report_buffer[index % 2], BUF_SIZE));
+		/*
+		 * Prevent reordering between the memcpy above and the atomic
+		 * read below.
+		 * See https://www.hpl.hp.com/techreports/2012/HPL-2012-68.pdf
```

Comment:
We could just refer to our own <include/linux/seqlock.h>, as this just mirrors that implementation. Hans's implementation is very C++11 specific, which doesn't quite apply in LKMM context.

Reply:
Done. I've also rewritten the loop to be more idiomatic.
All in all I decided to not use seqcount_latch_t, which basically just provides a wrapper around an atomic counter, but still requires the user to e.g. calculate the buffer index manually.

Reply:
That's fair. The most optimal implementation is the one reusing num_reports.

seqcount_latch_t takes care of all the memory-ordering bits, which is important during review. But we'll see if people dislike the open-coded seqlock or not.